src/cascadia/TerminalCore/TerminalApi.cpp

```diff
@@ -592,3 +594,55 @@ bool Terminal::EndHyperlink() noexcept
     _buffer->SetCurrentAttributes(attr);
     return true;
 }
+
+bool Terminal::SetWorkingDirectory(std::wstring_view uri) noexcept
+{
+    if (uri.size() < 8)
+    {
+        return false;
+    }
+
+    const auto prefix = uri.substr(0, 7);
+    if (!prefix.compare(L"file://") == 0)

```

Comment:
The [Uri class](https://docs.microsoft.com/en-us/uwp/api/Windows.Foundation.Uri?view=winrt-19041) might be useful here - you can use it to easily obtain the scheme/hostname/path etc

Reply:
I actually can not find a way to use `WIndows::Founcation::Uri` class. I have little experience with WinRT C++.

By the way do we need to worry about using this on Win7 or something?

Reply:
Nope. Windows Terminal needs [Windows 10 1903 or later](https://github.com/microsoft/terminal#installing-and-running-windows-terminal).

Reply:
The [WinRT introduction](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/intro-to-using-cpp-with-winrt#windows-desktop-application-cwinrt) has an example of specifically using `Windows.Foundation.Uri` from Win32 C++. It looks like the examples come with the WinRT VSIX extension.

The [C++ code itself looks reasonable](https://github.com/microsoft/cppwinrt/blob/master/vsix/ProjectTemplates/VC/Windows%20Desktop/ConsoleApplication/main.cpp), but if this is the first use of WinRT in the project, you might need to do other things to set up the include paths and linking. I've not done any development on Windows Terminal itself, so I don't know if that'll be simple like the [README implies](https://github.com/microsoft/cppwinrt/tree/master/vsix/ProjectTemplates/VC/Windows%20Desktop/ConsoleApplication), or complex if that README is hiding more complicated steps.

Reply:
The shared components like DX renderer actually require Win7 support, because it will be integrated into Visual Studio in the future. I'm not sure `TerminalApi` falls into that category.

Reply:
A quick poke around shows that [_TerminalCore/Terminal.cpp_](https://github.com/microsoft/terminal/blob/master/src/cascadia/TerminalCore/Terminal.cpp) is already using WinRT, for access to `TerminalControl`, so I _expect_ it would be as easy as
````
#include <winrt/Windows.Foundation.h>
```

and then using `::winrt::Windows::Foundation::Uri` per [the docs](https://docs.microsoft.com/en-us/uwp/api/Windows.Foundation.Uri?view=winrt-18362). (Make sure the docs are in C++/WinRT mode, it defaults to JavaScript mode for me for some reason)

Reply:
Despite the fact those WinRT goodies are very handy, I can not use it to parse the OSC 7 URI. The `::winrt::Windows::Foundation::Uri` class expect the URI format to be:

```plaintext
scheme://username:password @host:port/path.extension ?query #fragment
```

The FreeDesktop File URI spec is:

```plaintext
file://<hostname>/<path>
```

The behaviour of winrt Uri class is not very satisficing. For example for an Uri  `file:///D:/`, the winrt Uri gives you this:

```plaintext
Scheme: file
Host: 
Path: /D:/
```

Even weird, for an Uri in PowerShell `file://COMPUTERNAME/C:\\`, the winrt Uri class gives you this:

```plaintext
Scheme: file
Host: computername:
Path: /
```

I suspect that it can not handle `\` in a URI. And it prefers lowercase hostname? I don't really get it. We are, after all, parsing a FreeDesktop file URI so I can expect it to work. So that isn't a good choice to me.

Reply:
Those URI parses examples are basically correct. I don't know what you're expecting differently in the first case? Hostnames in URIs are not case-sensitive (leads back to the DNS spec, which is not case-sensitive).

For the second case, `\` is not a valid path separator in a URI, so I somewhat expected a path `/C:\\/` but I'm not overly surprised that it just fails weirdly, since that's not a valid Windows file:// URI (I don't believe `:` is valid in a UNC path name) and I assume `Windows::Foundation::Uri` is noticing that. I know `system.uri` in PowerShell does checking for `file://` URIs and rejects "invalid" ones (like `file://wsl$/...`, which is invalid-per-the-URI-spec, even though it's valid (surprisingly) in the Windows file:// URI spec, because Windows supports non-DNS-valid hostnames).

Reply:
That said, we should never see `\` in an OSC 7 URI, they need to be FreeDesktop File URI spec URIs (or as close as we can do) and the path separator is `/`, not `\`. So if we get `file:///D:/`, then the correct parse is as you gave above, which if we feed it back into urlmon, will be `D:\`.

Reply:
Even though I get the correct path, the Uri class is not happy with the host name. For a URI `file://COMPUTER-NAME/D:/Projects`, the Uri class gives you this:

```plaintext
Scheme: file
Host: computername:
Path: /D:/Projects
```

Sounds like if we are using the Uri class we'll have to give up on hostname, because it is in fact not a valid file URI on WIndows. The hostname MUST be removed.

Reply:
Interesting. Seems like a bug? `system.uri` handles that fine.
```
PS C:\Users\paulh> [system.uri]("file://COMPUTER-NAME/D:/Projects")

AbsolutePath   : /D:/Projects
AbsoluteUri    : file://computer-name/D:/Projects
LocalPath      : \\computer-name\D:\Projects
Authority      : computer-name
HostNameType   : Dns
IsDefaultPort  : True
IsFile         : True
IsLoopback     : False
PathAndQuery   : /D:/Projects
Segments       : {/, D:/, Projects}
IsUnc          : True
Host           : computer-name
Port           : -1
Query          :
Fragment       :
Scheme         : file
OriginalString : file://COMPUTER-NAME/D:/Projects
DnsSafeHost    : computer-name
IdnHost        : computer-name
IsAbsoluteUri  : True
UserEscaped    : False
UserInfo       :
```